### Youtube Anleitungen  
1) https://www.youtube.com/watch?v=HteW2lfwLXM  
2) https://www.youtube.com/watch?v=eLY8YAVnx_w

In [1]:
import pandas as pd
import ray
from ray import tune

import sys
import os
from dotenv import load_dotenv

# load the .env file
load_dotenv()
# append the absolute_project_path from .env variable to the sys.path
sys.path.append(os.environ.get('absolute_project_path'))

In [2]:
import os

python_path = os.environ.get('PYTHONPATH', '')
src_path = os.path.join(os.environ.get('absolute_project_path'), 'src')
os.environ['PYTHONPATH'] = f"{python_path}:{src_path}"
print(os.environ.get('PYTHONPATH'))

:/Users/mayte/GitHub/BugPlusEngine/src


In [3]:
# from src.environment import environment_ray as environment
from src.environment.environment_ray import BugPlus
from src.utils.matrix import number_bugs, array_to_matrices

In [4]:
import ray
from ray import tune
import ray.rllib.algorithms.ppo
import ray.rllib.algorithms.dqn

In [5]:
ray.init()

2023-03-02 13:44:45,576	INFO worker.py:1553 -- Started a local Ray instance.


RayContext(dashboard_url='', python_version='3.10.6', ray_version='2.3.0', ray_commit='cf7a56b4b0b648c324722df7c99c168e92ff0b45', address_info={'node_ip_address': '127.0.0.1', 'raylet_ip_address': '127.0.0.1', 'redis_address': None, 'object_store_address': '/tmp/ray/session_2023-03-02_13-44-43_100552_18131/sockets/plasma_store', 'raylet_socket_name': '/tmp/ray/session_2023-03-02_13-44-43_100552_18131/sockets/raylet', 'webui_url': '', 'session_dir': '/tmp/ray/session_2023-03-02_13-44-43_100552_18131', 'metrics_export_port': 62884, 'gcs_address': '127.0.0.1:59988', 'address': '127.0.0.1:59988', 'dashboard_agent_listen_port': 52365, 'node_id': 'fbdfd5ebd0d84c93d9b71381575df0473fd018b5694b7969d792aa6b'})

In [6]:
""" 
# configuration from the first ray try
config = (  # 1. Configure the algorithm,
    DQNConfig()
    .environment(environment.BugPlus)
    .rollouts(num_rollout_workers=2)
    .framework("tf2")
    .training(model={"fcnet_hiddens": [64, 64]},)
    .evaluation(evaluation_num_workers=1)
    .evaluation(evaluation_interval=500)
    # .spec(max_episode_steps=1)
)
"""

' \n# configuration from the first ray try\nconfig = (  # 1. Configure the algorithm,\n    DQNConfig()\n    .environment(environment.BugPlus)\n    .rollouts(num_rollout_workers=2)\n    .framework("tf2")\n    .training(model={"fcnet_hiddens": [64, 64]},)\n    .evaluation(evaluation_num_workers=1)\n    .evaluation(evaluation_interval=500)\n    # .spec(max_episode_steps=1)\n)\n'

In [8]:
tune.run("PPO",
         config = {"env": BugPlus
         })

2023-03-02 13:45:18,635	ERROR ray_trial_executor.py:930 -- An exception occurred when trying to stop the Ray actor:Traceback (most recent call last):
  File "/Users/mayte/GitHub/BugPlusEngine/venv/lib/python3.10/site-packages/ray/tune/execution/ray_trial_executor.py", line 921, in _resolve_stop_event
    ray.get(future, timeout=timeout)
  File "/Users/mayte/GitHub/BugPlusEngine/venv/lib/python3.10/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/Users/mayte/GitHub/BugPlusEngine/venv/lib/python3.10/site-packages/ray/_private/worker.py", line 2382, in get
    raise value
ray.exceptions.RayActorError: The actor died because of an error raised in its creation task, ray::PPO.__init__() (pid=18277, ip=127.0.0.1, repr=PPO)
  At least one of the input arguments for this task could not be computed:
ray.exceptions.RaySystemError: System error: No module named 'src'
traceback: Traceback (most recent call last):
ModuleNotFoundError: No

TuneError: ('Trials did not complete', [PPO_BugPlus_11ac2_00000])

(PPO pid=18277) 2023-03-02 13:45:18,612	ERROR serialization.py:371 -- No module named 'src'
(PPO pid=18277) Traceback (most recent call last):
(PPO pid=18277)   File "/Users/mayte/GitHub/BugPlusEngine/venv/lib/python3.10/site-packages/ray/_private/serialization.py", line 369, in deserialize_objects
(PPO pid=18277)     obj = self._deserialize_object(data, metadata, object_ref)
(PPO pid=18277)   File "/Users/mayte/GitHub/BugPlusEngine/venv/lib/python3.10/site-packages/ray/_private/serialization.py", line 252, in _deserialize_object
(PPO pid=18277)     return self._deserialize_msgpack_data(data, metadata_fields)
(PPO pid=18277)   File "/Users/mayte/GitHub/BugPlusEngine/venv/lib/python3.10/site-packages/ray/_private/serialization.py", line 207, in _deserialize_msgpack_data
(PPO pid=18277)     python_objects = self._deserialize_pickle5_data(pickle5_data)
(PPO pid=18277)   File "/Users/mayte/GitHub/BugPlusEngine/venv/lib/python3.10/site-packages/ray/_private/serialization.py", line 197, in _

2023-03-02 11:00:36,985	INFO worker.py:1553 -- Started a local Ray instance.
Ray
Python version:	3.10.6
Ray version:	2.3.0
2023-03-02 11:08:18,826	INFO algorithm_config.py:2888 -- Executing eagerly (framework='tf2'), with eager_tracing=tf2. For production workloads, make sure to set eager_tracing=True  in order to match the speed of tf-static-graph (framework='tf'). For debugging purposes, `eager_tracing=False` is the best choice.
2023-03-02 11:08:18,828	INFO algorithm_config.py:2888 -- Executing eagerly (framework='tf2'), with eager_tracing=tf2. For production workloads, make sure to set eager_tracing=True  in order to match the speed of tf-static-graph (framework='tf'). For debugging purposes, `eager_tracing=False` is the best choice.